core/arch/aarchxx/mangle.c

```diff
@@ -2922,7 +2954,573 @@ mangle_icache_op(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+#else
+        /* Pair store requires consecutive register numbers. */
+        PRE(ilist, where,
+            instr_create_save_to_tls(dcontext, base_reg, TLS_LDSTEX_ADDR_SLOT));
+        PRE(ilist, where,
+            instr_create_save_to_tls(dcontext, value_reg, TLS_LDSTEX_VALUE_SLOT));
+#endif
+        /* Write the size to TLS, using a scratch register. */
+        ushort slot;
+        bool should_restore;
+        reg_id_t scratch = pick_scratch_reg(dcontext, instr, swap_reg, DR_REG_NULL,
+                                            DR_REG_NULL, true, &slot, &should_restore);
+        insert_save_to_tls_if_necessary(dcontext, ilist, where, scratch, slot);
+        PRE(ilist, where,
+            XINST_CREATE_load_int(
+                dcontext, opnd_create_reg(scratch),
+                OPND_CREATE_INT(opnd_get_size(instr_get_dst(instr, 0)))));
+        /* If a load-pair, write the 2nd value as well. */
+        if (instr_num_dsts(instr) == 2) {
+#ifdef AARCH64
+            PRE(ilist, where,
+                XINST_CREATE_store_pair(
+                    dcontext,
+                    opnd_create_base_disp(dr_reg_stolen, DR_REG_NULL, 0,
+                                          TLS_LDSTEX_VALUE2_SLOT,
+                                          IF_ARM_ELSE(OPSZ_8, OPSZ_16)),
+                    opnd_create_reg(value2_reg), opnd_create_reg(scratch)));
+#else
+            /* Pair store requires consecutive register numbers.
+             * XXX: We could store the 2 values at once.
+             */
+            PRE(ilist, where,
+                instr_create_save_to_tls(dcontext, value2_reg, TLS_LDSTEX_VALUE2_SLOT));
+            PRE(ilist, where,
+                instr_create_save_to_tls(dcontext, scratch, TLS_LDSTEX_SIZE_SLOT));
+#endif
+        } else {
+            PRE(ilist, where,
+                instr_create_save_to_tls(dcontext, scratch, TLS_LDSTEX_SIZE_SLOT));
+        }
+        if (should_restore) {
+            PRE(ilist, where, instr_create_restore_from_tls(dcontext, scratch, slot));
+        }
+    } else {
+        /* Insert a label so subsequent stex mangling knows the ldex was here. */
+        instr_t *marker = INSTR_CREATE_label(dcontext);
+        instr_set_note(marker, (void *)DR_NOTE_LDEX);
+        dr_instr_label_data_t *label_data = instr_get_label_data_area(marker);
+        label_data->data[0] = value_reg;  /* Ruled out dr_reg_stolen above. */
+        label_data->data[1] = value2_reg; /* Ruled out dr_reg_stolen above. */
+        PRE(ilist, where, marker);
+    }
+    /* Replace the exclusive load with a non-exclusive version. */
+    instr_t *ld_nonex = create_ld_from_ldex(dcontext, instr);
+    instrlist_preinsert(ilist, instr, ld_nonex);
+    if (swap_reg != DR_REG_NULL) {
+        if (instr_writes_to_reg(instr, swap_reg, DR_QUERY_INCLUDE_COND_DSTS)) {
+            PRE(ilist, where,
+                instr_create_save_to_tls(dcontext, swap_reg, TLS_REG_STOLEN_SLOT));
+        }
+        if (swap_restore) {
```

Comment:
> Checking my understanding of this part:
> For the case where there was an exclusive load into the stolen reg, here we save the loaded value (which is in swap_reg because we replaced the stolen reg with swap_reg) to TLS_REG_STOLEN_SLOT, and then possibly restore the swap_reg's old value.

Yes, correct.

> Q: We probably also need to set the stolen_reg's to swap_reg's right? (For subsequent usages of the loaded value by app instrs)

I'm not sure I understand this as a separate question, because preserving the stolen register app value is  what is already done: that's what you already described as "save the loaded value to TLS_REG_STOLEN_SLOT".  How is your question here different?

Reply:
L3297 saves the loaded value (which is in `swap_reg`) to `TLS_REG_STOLEN_SLOT`. But where do we restore it (copy from the spill slot to the actual stolen reg). We'd need to restore it as subsequent app instrs expect the loaded val to be in the stolen reg.

Reply:
Still not understanding: the stolen reg always holds the TLS base address.  The app's actual stolen reg value is kept in TLS: it is never in the stolen reg itself (except when we mangle non-exclusive-load-store app instrs that use the stolen reg).

Reply:
Ah right. DR would restore stolen reg value around each app use.